xtuner/v1/profiler/prober.py

```diff
@@ -0,0 +1,915 @@
+import functools
+import json
+import re
+import sys
+import time
+from abc import ABC
+from pathlib import Path
+from typing import TYPE_CHECKING, Callable, ClassVar
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.tensor import DTensor
+
+from xtuner.v1.utils import get_logger
+
+
+if TYPE_CHECKING:
+    # Don't import model and module here to avoid circular import.
+    # Because the model and module may import ProberList to debug.
+    from xtuner.v1.module.decoder_layer.moe_decoder_layer import RouterResults
+else:
+    RouterResults = dict
+
+
+logger = get_logger()
+
+
+def get_dtensor_meta(dtensor: torch.Tensor):
+    if not isinstance(dtensor, DTensor):
+        return {}
+
+    return {
+        "local_shape": dtensor._local_tensor.shape,
+        "device_mesh": str(dtensor.device_mesh),
+        "placements": str(dtensor.placements),
+    }
+
+
+class BaseProber(ABC):
+    """
```

Comment:
English docstring for rendering the api docs. Besides, the layout of the docstring should be:

"""

[description]
[args]
"""

Summary should not be empty